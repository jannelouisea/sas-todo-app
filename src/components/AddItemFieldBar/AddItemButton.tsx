import IconButton from '@mui/joy/IconButton';
import Tooltip from '@mui/joy/Tooltip';
import AddIcon from '@mui/icons-material/Add';
import PropTypes from 'prop-types';

import {
    MUISize,
    MUIVariant,
    MUIPos,
} from 'src/static/enums';

type Props = {
    onClick: React.MouseEventHandler<HTMLButtonElement>
};

function AddItemButton({ onClick }: Props) {
    const muiButtonStyles = { borderRadius: '50%' };

    return (
        <Tooltip title='Add to-do item' size={MUISize.Medium} variant={MUIVariant.Outlined} placement={MUIPos.Right}>
            <IconButton
                id='add-item-button'
                className='drop-shadow'
                variant={MUIVariant.Solid}
                aria-label='Add to-do item button'
                size={MUISize.Large}
                onClick={onClick}
                sx={muiButtonStyles}
            >
                <AddIcon />
            </IconButton>
        </Tooltip>
    );
}

AddItemButton.propTypes = {
    onClick: PropTypes.object
}

export default AddItemButton;