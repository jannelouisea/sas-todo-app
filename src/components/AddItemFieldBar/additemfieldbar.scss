.add-item-field-bar {
    display: flex;

    .add-item-input {
        flex: 1 1 0%;
    }

    .add-item-button {
        margin-left: .5rem;
    }
}