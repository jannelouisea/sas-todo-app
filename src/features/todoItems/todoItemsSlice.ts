import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import {
    IToDoItem
} from 'src/interfaces'
import moment from 'moment';
import {
    Sort
} from 'src/static/enums'

export type ITodoItemAddedPayload = {
    id: string,
    text: string,
    createdAt: number,
};

const initialState: IToDoItem[] = [
    {
        id: 'init-item-001',
        text: 'Clean the house',
        createdAt: moment().unix(),
        completed: false
    },
    {
        id: 'init-item-002',
        text: 'Pack things for upcoming trip',
        createdAt: 1689652800, // Tue Jul 18 2023 00:00:00 GMT-0400 (EST)
        completed: false
    },
    {
        id: 'init-item-003',
        text: 'Make dinner',
        createdAt: 1690257600, // Tue Jul 25 2023 00:00:00 GMT-0400 (EST)
        completed: true
    },
    {
        id: 'init-item-004',
        text: 'Prepare for SAS Interview',
        createdAt: 1689393600, // Sat Jul 15 2023 00:00:00 GMT-0400 (EST)

        completed: true
    },
];

const todoItemsSlice = createSlice({
    name: 'todoItems',
    initialState,
    reducers: {
        todoItemAdded(state, action: PayloadAction<ITodoItemAddedPayload>) {
            const { id, text, createdAt } = action.payload;
            const newItem = {
                id,
                text,
                createdAt,
                completed: false
            };

            return [newItem, ...state];
        },
        todoItemToggled(state, action: PayloadAction<string>) {
            return state.map(item => {
                if (item.id !== action.payload) {
                    return item
                }

                return {
                    ...item,
                    completed: !item.completed
                }
            });
        },
        todoItemRemoved(state, action: PayloadAction<string>) {
            return state.filter(item => item.id !== action.payload);
        },
        todoItemsSorted(state, action: PayloadAction<Sort>) {
            let newState = [...state];

            switch (action.payload) {
                case Sort.DateDesc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemB.createdAt - itemA.createdAt
                        //moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                    );
                case Sort.DateAsc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemA.createdAt - itemB.createdAt
                        //moment(itemA.createdAt).format('YYYYMMDD').localeCompare(moment(itemB.createdAt).format('YYYYMMDD'))
                    );
                case Sort.TextDesc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemB.text.localeCompare(itemA.text)
                    );
                case Sort.TextAsc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemA.text.localeCompare(itemB.text)
                    );
                case Sort.Cmpl1st:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) => {
                        if (itemA.completed === itemB.completed) {
                            //return moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                            return itemB.createdAt - itemA.createdAt;
                        } else if (itemA.completed) {
                            return -1;
                        } else {
                            return 1;
                        }
                    });
                case Sort.Incmpl1st:
                default:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) => {
                        if (itemA.completed === itemB.completed) {
                            //return moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                            return itemB.createdAt - itemA.createdAt;
                        } else if (itemA.completed) {
                            return 1;
                        } else {
                            return -1;
                        }
                    });
            }
        }
    }
});

export const { todoItemAdded, todoItemToggled, todoItemRemoved, todoItemsSorted } = todoItemsSlice.actions;

export default todoItemsSlice.reducer;