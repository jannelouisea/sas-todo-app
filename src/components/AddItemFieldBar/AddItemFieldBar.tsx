import { useState, useEffect } from 'react';
import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { todoItemAdded, todoItemsSorted } from 'src/features/todoItems/todoItemsSlice';

import moment from 'moment';
import uniqid from 'uniqid';

import Input from '@mui/joy/Input';
import Typography from '@mui/joy/Typography';
import { AddItemButton } from 'src/components';

import {
    MUISize,
    MUIVariant,
    MUIText,
    MUIColor,
} from 'src/static/enums';

function AddItemFieldBar() {
    const INITIAL_TEXT_STATE = '';
    const IS_TEXT_EMPTY_ERR_MSG = 'To-do item cannot be blank.';
    const PLACEHOLDER = 'What do I need to-do?';

    const dispatch = useAppDispatch();
    const sortBy = useAppSelector(state => state.todoItemsFilters.sortBy);

    const [text, setText] = useState(INITIAL_TEXT_STATE);
    const [textChanged, setTextChanged] = useState(false);
    const [hasError, setHasError] = useState(false);

    useEffect(() => {
        const isEmpty = textChanged && text === '';
        setHasError(isEmpty);
    }, [text, textChanged]);

    const onTextChanged = (evt: React.ChangeEvent<HTMLInputElement>) => {
        setTextChanged(true);
        setText(evt.target.value);
    }

    const onAddItem = () => {
        if (text !== '') {
            dispatch(todoItemAdded({
                id: uniqid(),
                text,
                createdAt: moment().toDate(),
            }));
            dispatch(todoItemsSorted(sortBy));
            resetState();
        } else {
            setHasError(true);
        }
    };

    const onInputBlur = () => {
        setTextChanged(false);
    }

    const resetState = () => {
        setTextChanged(false);
        setText(INITIAL_TEXT_STATE);
    };

    const muiInputStyles = { borderRadius: '4rem' };
    const muiErrStyles = { marginTop: '.25rem', paddingLeft: '1.5rem' };

    return (
        <div className='add-item-field-bar'>
            <div className='add-item-input'>
                <Input
                    className='drop-shadow'
                    placeholder={PLACEHOLDER}
                    size={MUISize.Large}
                    value={text}
                    onChange={onTextChanged}
                    onBlur={onInputBlur}
                    sx={muiInputStyles}
                    error={hasError}
                    aria-label='Add to-do item input field'
                />
                {
                    hasError &&
                    <Typography
                        color={MUIColor.Danger}
                        level={MUIText.Label}
                        variant={MUIVariant.Plain}
                        sx={muiErrStyles}
                    >
                        {IS_TEXT_EMPTY_ERR_MSG}
                    </Typography>
                }
            </div>
            <div className='add-item-button'>
                <AddItemButton onClick={onAddItem} />
            </div>
        </div>
    );
}

export default AddItemFieldBar;