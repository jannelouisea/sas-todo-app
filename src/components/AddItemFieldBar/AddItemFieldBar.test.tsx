import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import * as todoItemsActions from 'src/features/todoItems/todoItemsSlice';

import {
    IStoreType,
    createStore
} from 'src/app/store'

import AddItemFieldBar from './AddItemFieldBar';

describe('AddItemFieldBar', () => {
    const inputLabelTextId = 'Add to-do item input field';
    const addButtonLabelTextId = 'Add to-do item button';
    const IS_TEXT_EMPTY_ERR_MSG = 'To-do item cannot be blank.';

    let store: IStoreType;

    const errMsgNotFound = () => {
        let errFound = true;
        try {
            screen.getByText(IS_TEXT_EMPTY_ERR_MSG);
        } catch (e) {
            errFound = false;
        } finally {
            expect(errFound).toBeFalsy();
        }
    };

    beforeEach(() => {
        store = createStore();
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <AddItemFieldBar />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('adds a new to-do item', () => {
        const todoItemAddedSpy = jest.spyOn(todoItemsActions, 'todoItemAdded');
        const todoItemsSortedSpy = jest.spyOn(todoItemsActions, 'todoItemsSorted');

        render(
            <Provider store={store}>
                <AddItemFieldBar />
            </Provider>
        );

        // Verify entering to-do item text
        let input: HTMLInputElement = screen.getByLabelText(inputLabelTextId);
        expect(input).toBeInTheDocument();

        const todoItemText = 'Clean dishes';
        fireEvent.change(input, { target: { value: todoItemText } });
        input = screen.getByLabelText(inputLabelTextId);
        expect(input.value).toEqual(todoItemText);

        // Verify adding to-do item to store
        let addButton: HTMLInputElement = screen.getByLabelText(addButtonLabelTextId);
        expect(addButton).toBeInTheDocument();

        fireEvent.click(addButton);

        expect(todoItemAddedSpy).toHaveBeenCalledTimes(1);
        expect(todoItemsSortedSpy).toHaveBeenCalledTimes(1);

        errMsgNotFound();
    });

    it('does not add new to-do item when text is empty', () => {
        const todoItemAddedSpy = jest.spyOn(todoItemsActions, 'todoItemAdded');
        const todoItemsSortedSpy = jest.spyOn(todoItemsActions, 'todoItemsSorted');

        render(
            <Provider store={store}>
                <AddItemFieldBar />
            </Provider>
        );

        // Verify entering to-do item text
        let input: HTMLInputElement = screen.getByLabelText(inputLabelTextId);
        expect(input).toBeInTheDocument();

        const todoItemText = '';
        fireEvent.change(input, { target: { value: todoItemText } });
        input = screen.getByLabelText(inputLabelTextId);
        expect(input.value).toEqual(todoItemText);

        // Verify adding to-do item to store
        let addButton: HTMLInputElement = screen.getByLabelText(addButtonLabelTextId);
        expect(addButton).toBeInTheDocument();

        fireEvent.click(addButton);

        expect(todoItemAddedSpy).toHaveBeenCalledTimes(0);
        expect(todoItemsSortedSpy).toHaveBeenCalledTimes(0);

        // Verify error message
        let errMsg = screen.getByText(IS_TEXT_EMPTY_ERR_MSG);
        expect(errMsg).toBeInTheDocument();
    });

    it('shows error message when user enters blank item text', () => {
        render(
            <Provider store={store}>
                <AddItemFieldBar />
            </Provider>
        );

        // Verify entering to-do item text
        let input: HTMLInputElement = screen.getByLabelText(inputLabelTextId);
        expect(input).toBeInTheDocument();

        let todoItemText = 'not empty';
        fireEvent.change(input, { target: { value: todoItemText } });
        input = screen.getByLabelText(inputLabelTextId);
        expect(input.value).toEqual(todoItemText);

        errMsgNotFound();

        todoItemText = '';
        fireEvent.change(input, { target: { value: todoItemText } });
        input = screen.getByLabelText(inputLabelTextId);
        expect(input.value).toEqual(todoItemText);

        // Verify error message
        let errMsg = screen.getByText(IS_TEXT_EMPTY_ERR_MSG);
        expect(errMsg).toBeInTheDocument();
    });
})