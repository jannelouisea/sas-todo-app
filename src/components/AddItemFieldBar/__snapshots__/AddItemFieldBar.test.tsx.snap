// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItemFieldBar renders correctly 1`] = `
<div
  className="add-item-field-bar"
>
  <div
    className="add-item-input"
  >
    <div
      className="MuiInput-root MuiInput-variantOutlined MuiInput-colorNeutral MuiInput-sizeLg drop-shadow css-1n5ydwi-JoyInput-root"
      onClick={[Function]}
    >
      <input
        aria-label="Add to-do item input field"
        className="MuiInput-input css-1gw9vc6-JoyInput-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="What do I need to-do?"
        required={false}
        value=""
      />
    </div>
  </div>
  <div
    className="add-item-button"
  >
    <button
      aria-label="Add to-do item button"
      aria-labelledby={null}
      className="MuiIconButton-root MuiIconButton-variantSolid MuiIconButton-colorPrimary MuiIconButton-sizeLg drop-shadow css-1tp8nsa-JoyIconButton-root"
      id="add-item-button"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </button>
  </div>
</div>
`;
